File: src/org/simbrain/world/imageworld/ImageWorldComponent.java
package org.simbrain.world.imageworld;

import com.thoughtworks.xstream.XStream;
import com.thoughtworks.xstream.io.xml.DomDriver;
import org.simbrain.workspace.AttributeContainer;
import org.simbrain.workspace.WorkspaceComponent;
import org.simbrain.world.imageworld.serialization.BufferedImageConverter;
import org.simbrain.world.imageworld.serialization.CouplingArrayConverter;

import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.List;

/**
 * ImageWorldComponent provides a model for building an image processing
 * pipeline and coupling inputs and outputs within a Simbrain workspace.
 *
 * @author Tim Shea
 */
public abstract class ImageWorldComponent extends WorkspaceComponent {

    public ImageWorldComponent() {
        super("");
    }

    public ImageWorldComponent(String name) {
        super(name);
    }

    /**
     * Create an xstream from this class.
     */
    public static XStream getXStream() {
        XStream stream = new XStream(new DomDriver());
        stream.registerConverter(new BufferedImageConverter());
        stream.registerConverter(new CouplingArrayConverter());
        return stream;
    }

    @Override
    public void save(OutputStream output, String format) {
        getXStream().toXML(getWorld(), output);
    }

    @Override
    protected void closing() {
    }

    public abstract ImageWorld getWorld();

    @Override
    public List<AttributeContainer> getAttributeContainers() {
        List<AttributeContainer> models = new ArrayList<>();
        models.addAll(getWorld().getSensorMatrices());
        models.addAll(getWorld().getImageSources());
        return models;
    }

    @Override
    public AttributeContainer getObjectFromKey(String objectKey) {
        for (ImageSource source : getWorld().getImageSources()) {
            if (objectKey.equals(source.getClass().getSimpleName())) {
                return source;
            }
        }
        for (SensorMatrix sensor : getWorld().getSensorMatrices()) {
            if (objectKey.equals(sensor.getName())) {
                return sensor;
            }
        }
        return null;
    }

    @Override
    public void update() {
        getWorld().update();
    }

}

File: src/main/kotlin/org/simbrain/world/imageworld/ImageWorldComponent.kt
package org.simbrain.world.imageworld

import com.thoughtworks.xstream.XStream
import org.simbrain.util.getSimbrainXStream
import org.simbrain.workspace.AttributeContainer
import org.simbrain.workspace.WorkspaceComponent
import org.simbrain.world.imageworld.serialization.BufferedImageConverter
import org.simbrain.world.imageworld.serialization.CouplingArrayConverter
import java.io.InputStream
import java.io.OutputStream

/**
 * The interface between image world and the desktop level.
 * Manages couplings and persistence.
 */
class ImageWorldComponent : WorkspaceComponent {
    /**
     * The image world this component displays.
     */
    val world: ImageWorld

    constructor() : super("") {
        this.world = ImageWorld()
    }

    /**
     * Create named component.
     */
    constructor(name: String) : super(name) {
        this.world = ImageWorld()
    }

    override val attributeContainers: List<AttributeContainer>
        get() = buildList {
            add(world.imageAlbum)
            add(world.imagePipelineCollection)
            addAll(world.imagePipelineCollection.pipelines)
        }

    override val xml: String?
        get() = xStream.toXML(world)

    override fun save(output: OutputStream, format: String?) {
        xStream.toXML(world, output)
    }

    /**
     * Deserialize an ImageAlbumComponent.
     *
     * @param name name of component
     * @param world the deserialized world
     */
    constructor(name: String, world: ImageWorld) : super(name) {
        this.world = world
    }

    companion object {
        /**
         * Open a saved ImageWorldComponent from an XML input stream.
         *
         * @param input  The input stream to read.
         * @param name   The name of the new world component.
         * @param format The format of the input stream. Should be xml.
         * @return A deserialized ImageWorldComponent.
         */
        fun open(input: InputStream?, name: String, format: String?): ImageWorldComponent {
            val world = xStream.fromXML(input) as ImageWorld
            return ImageWorldComponent(name, world)
        }

        val xStream: XStream
            /**
             * Create an xstream from this class.
             */
            get() {
                val stream = getSimbrainXStream()
                stream.registerConverter(BufferedImageConverter())
                stream.registerConverter(CouplingArrayConverter())
                return stream
            }
    }
}